frb/analysis/kcwi.py

```diff
@@ -113,358 +66,473 @@ def silence_warnings(warncategory):
-
-    wave = cube.spectral_axis.value
-
-    # Convert to vacuum wavelengths?
-    if tovac:
-        wave = _air_to_vac(wave)
-    # Prepare an HDU in advance
-    wavehdu = fits.ImageHDU(wave)
-    wavehdu.header.set('extname', 'WAVELENGTH')
-
-    # Initialize output lists 
-    speclist = np.zeros([nobjs, len(wave)])
-    varspeclist = np.zeros_like(speclist)
-
-    # Create output folder?
-    if not os.path.isdir(outdir):
-        os.mkdir(outdir)
-    for idx, obj in enumerate(objects):
-        spec, varspec = spec_from_ellipse(cube, varcube,
-                                          obj['x'], obj['y'],
-                                          obj['a'], obj['b'],
-                                          obj['theta'], r = 2)
-
-        # Produce spectrum fits file
-        spechdu = fits.PrimaryHDU(spec.data, header=spec.header)
-        spechdu.header.set('extname', 'SPEC')
-        varhdu = fits.ImageHDU(varspec.data, header=varspec.header)
-        varhdu.header.set('extname', 'VAR')
-        hdulist = fits.HDUList([spechdu, varhdu, wavehdu])
-        specfile_name = outdir+str(idx)+"_spec1d.fits"
-        hdulist.writeto(specfile_name, overwrite=True)
-
-        # Append spectrum to list
-        speclist[idx] = spec.data
-        varspeclist[idx] = varspec.data
-
-    if marzfile:
-        _make_marz(cube, speclist, varspeclist, objects, outdir+marzfile, tovac=tovac)
-    return speclist, varspeclist, wave
+        # if kind is max:
+        # if kind is something_else:
+        masked_var = self.varcube.subcube_from_mask(mask_arr)
+        varspec = masked_var.mean(axis = (1,2), how = how)/np.sum(mask_arr) # Var of the mean = Sum of var/npix^2 = Mean of var/npix assuming each pix is uncorelated (not true, really).
+        return spec, varspec
+
+    def spec_from_ellipse(self,
+                        x0 = 0., y0 = 0., a = 1.,
+                        b = 1., theta = 0.):
+        
+        """
+        Get the spectrum within an elliptical region
+        Args:
+            cube (Spectral Cube): A datacube object
+            x0, y0 (float, optional): Centroid of ellipse
+            a, b (float, optional): semi-major and semi-minor axes
+            theta (float, optional): rotation angle of the semi-major
+                axis from the positive x axis.
+        Returns:
+            spec (OneDSpectrum): The extracted spectrum.
+            var (OneDSpectrum): Variance in spectrum.
+                Only returned if varcube is supplied.
+        """
+        #TODO: Use photutils aperture object for this.
+        # Create 2D mask first
+        ellipse = EllipticalAperture((x0, y0), a, b, theta)
+        # Create a mask from the ellipse
+        aper_mask = ellipse.to_mask(method='center').to_image(self.cube.shape[1:]).astype(bool)
+        
+        return self.spec_from_spatial_mask(aper_mask)
+
+
+    def _make_marz(self, speclist, varspeclist, objects,  wave=None, marzfile="marzfile.fits", tovac = True):
```

Comment:
why is varspeclist not calculated internally from self.varcube?

Reply:
It is. The specrtra and varspecs are being produced by spec from spatial mask. This function only takes a list of those calculated values and formats them into a MARZ file.